designs/v1-roadmap.md

```diff
@@ -0,0 +1,203 @@
+
+We need to implement a common way for handing breaking changes to our hashing logic ahead of v1 so we can make the requisite changes that we need to make to the v1 APIs at v1 as well as handle breaking changes through defaults to our drift hashing API moving forward post-v1 e.g. we introduce an alpha field that affects our static hashing and then drop it later.
+
+#### Tasks
+
+- [ ] Design and implement a hash version-style implementation that allows us to make breaking changes to the hash versioning scheme
+
+### Introduce ConsolidateAfter for Consolidation Controls
+
+**Issue Ref(s):** https://github.com/kubernetes-sigs/karpenter/issues/735
+
+**Category**: Breaking, Stability
+
+Karpenter currently evaluates all forms of disruption in a synchronous loop, starting with expiration, drift, emptiness, and considering consolidation and multi-node consolidation *only* if the other conditions are not satisfied. Consolidation performs scheduling simulations on the cluster and evaluates if there are any opportunities to save money by removing nodes or replacing nodes on the cluster.
+
+The current consolidateAfter behavior creates a consolidation decision, waits for 15s synchronously inside of the disruption loop, and then re-validates that the same decision is still valid. This was intended to address the concern from [users that consolidation was acting too aggressively](https://github.com/aws/karpenter-provider-aws/issues/2370), and consolidating nodes that, if kept around for a little longer, would be valid for pods to schedule to. Adding the synchronous wait had the desired effect, but it caused us to have to keep this value low since it blocks *all* forms of disruption while it is waiting.
+
+Users have asked that we make this a configurable field so that they can tweak whether we keep nodes around for longer when they are underutilized before we make a consolidation decision to terminate them. To do this, we will have to change our synchronous waiting mechanism to some other mechanism that will allow us to perform longer waits.
+
+#### Tasks
+
+- [ ] Design and implement a `spec.consolidateAfter` field for the v1 NodePool API, reworking our synchronous wait to ensure that waiting for nodes that haven’t reached the end of their `consolidateAfter` timeframe doesn’t block other disruption evaluation
+
+### Change Expiration from Graceful to Forceful
+
+**Issue/RFC Ref:** https://github.com/kubernetes-sigs/karpenter/pull/1303
+
+**Category:** Breaking
+
+Users care about the removal of their nodes within a certain timeframe due to security requirements. Budgets, `karpenter.sh/do-not-disrupt` pods and blocking PDBs should not stop us from starting the drain of the node after it exceeds its expireAfter value. Once a node has begun evicting, it will remain on the cluster up to its t`erminationGracePeriod` duration (assuming there are blocking PDBs or pods that are running with the `karpenter.sh/do-not-disrupt` annotation) and then be removed.
```

Comment:
```suggestion
Users care about the removal of their nodes within a certain timeframe due to security requirements. Budgets, `karpenter.sh/do-not-disrupt` pods and blocking PDBs should not stop us from starting the drain of the node after it exceeds its expireAfter value. Once a node has begun evicting, it will remain on the cluster up to its `terminationGracePeriod` duration (assuming there are blocking PDBs or pods that are running with the `karpenter.sh/do-not-disrupt` annotation) and then be removed.
```